src/collision_detection/fcl/fcl_model.cpp

```diff
@@ -236,28 +236,159 @@ void FCLModelTpl<S>::updateCollisionObjects(
 }
 
 template <typename S>
-bool FCLModelTpl<S>::collide(const fcl::CollisionRequest<S> &request) const {
-  fcl::CollisionResult<S> result;
-  for (const auto &[i, j] : collision_pairs_) {
-    collision_detection::fcl::collide(collision_objects_[i], collision_objects_[j],
-                                      request, result);
-    if (result.isCollision()) return true;
-  }
-  return false;
+std::vector<WorldCollisionResultTpl<S>> FCLModelTpl<S>::checkSelfCollision(
+    const CollisionRequest &request, const AllowedCollisionMatrixPtr &acm) const {
+  std::vector<WorldCollisionResult> ret;
+  CollisionResult result;
+
+  for (const auto &[i, j] : collision_pairs_)
+    if (auto type = acm->getAllowedCollision(collision_link_names_[i],
+                                             collision_link_names_[j]);
+        !type || type == collision_detection::AllowedCollision::NEVER) {
+      result.clear();
+      collision_detection::fcl::collide(collision_objects_[i], collision_objects_[j],
+                                        request, result);
+      if (result.isCollision()) {
+        WorldCollisionResult tmp;
+        tmp.res = result;
+        tmp.collision_type = "self";
+        tmp.object_name1 = name_;
+        tmp.object_name2 = name_;
+        tmp.link_name1 = collision_link_names_[i];
+        tmp.link_name2 = collision_link_names_[j];
+        ret.push_back(tmp);
+      }
+    }
+  return ret;
 }
 
 template <typename S>
-std::vector<fcl::CollisionResult<S>> FCLModelTpl<S>::collideFull(
-    const fcl::CollisionRequest<S> &request) const {
-  // TODO(merge): return only CollisionResult in collision?
-  // Result will be returned via the collision result structure
-  std::vector<fcl::CollisionResult<S>> ret;
-  for (const auto &[i, j] : collision_pairs_) {
-    fcl::CollisionResult<S> result;
-    collision_detection::fcl::collide(collision_objects_[i], collision_objects_[j],
-                                      request, result);
-    ret.push_back(result);
-  }
+std::vector<WorldCollisionResultTpl<S>> FCLModelTpl<S>::checkCollisionWith(
+    const FCLModelTplPtr<S> &other, const CollisionRequest &request,
+    const AllowedCollisionMatrixPtr &acm) const {
+  std::vector<WorldCollisionResult> ret;
+  CollisionResult result;
+
+  for (const auto &col_obj : collision_objects_)
+    for (const auto &col_obj2 : other->collision_objects_)
+      if (auto type = acm->getAllowedCollision(col_obj->name, col_obj2->name);
```

Comment:
This might be a bug since we are checking collision across two different models with the same link names. Say if model 1 has collision disabled for link 1 and link 5. Model 2 happens to have link 5, then we will not try to collide model 1's link 1 and model 2's link 5

Reply:
This requires adding unique link_names in `mplib.sapien_utils.conversion.convert_object_name()` and use this to get link names in srdf_exporter/urdf_exporter